### ISSUE TITLE
Elixir 1.14.1: defimpl ..., for: __MODULE__ does not work, worked correctly in 1.14.0

### ISSUE THREAD

**Issue by jonklein** [elixir-lang/elixir#12216]
### Elixir and Erlang/OTP versions

```
% elixir --version
Erlang/OTP 25 [erts-13.1.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit] [dtrace]

Elixir 1.14.0 (compiled with Erlang/OTP 23)
```

### Operating system

Mac OS X 12.6 (Darwin 21.6.0 Darwin Kernel Version 21.6.0: Mon Aug 22 20:19:52 PDT 2022; root:xnu-8020.140.49~2/RELEASE_ARM64_T6000 arm64)

### Current behavior

Using `__MODULE__`  with `defimpl` is not working as expected.  Using the module name explicitly does work.

```
# impl.ex
defprotocol Proto do
  @spec type(t) :: String.t()
  def type(value)
end

defmodule A do
  defstruct value: ""

  defimpl Proto, for: __MODULE__ do
    def type(_value), do: "A"
  end
end
```

```
# impl_text.ex
IO.inspect(Proto.type(%A{}))
```

```
% asdf local elixir 1.14.1
% elixir -r impl.ex impl_test.ex
** (Protocol.UndefinedError) protocol Proto not implemented for %A{value: ""} of type A (a struct)
    impl.ex:1: Proto.impl_for!/1
    impl.ex:3: Proto.type/1
    impl_test.ex:1: (file)
```

### Expected behavior

The same code works as expected in Elixir 1.14.0.  I see nothing in the CHANGELOG to indicate this would be an expected behavior.

```
% asdf local elixir 1.14.0
% elixir -r impl.ex impl_test.ex
"A"
```

**Comment by josevalim** [comment]
Yes, it is a bug and it will be addressed in v1.14.1. Note you can remove `for: __MODULE__` altogether in said cases and it will work since Elixir v1.0.

**Comment by josevalim** [comment]
Duplicate of #12186.
